packages/serverpod/lib/src/server/websocket_request_handlers/method_websocket_request_handler.dart

```diff
@@ -172,95 +212,292 @@ class MethodWebsocketRequestHandler {
-        endpointMethodConnector,
-        session,
-        args,
-        message,
-        server,
-      );
-    } else if (endpointMethodConnector is MethodConnector) {
-      _handleMethodCallEndpoint(
-        endpointMethodConnector,
-        session,
-        args,
-        message,
-        server,
-      );
+    _registerOutputStream(webSocket, message, methodConnector);
+    var inputStreams = _createInputStreams(
+      methodConnector,
+      webSocket,
+      message,
+    );
+    var streamParams = inputStreams.map(
+      (key, value) => MapEntry(key, value.stream),
+    );
+
+    switch (methodConnector.returnType) {
+      case MethodStreamReturnType.streamType:
+        _handleMethodWithStreamReturn(
+          methodConnector: methodConnector,
+          session: session,
+          args: args,
+          streamParams: streamParams,
+          message: message,
+          server: server,
+        );
+        break;
+      case MethodStreamReturnType.singleType:
+      case MethodStreamReturnType.voidType:
+        _handleMethodWithFutureReturn(
+          methodConnector: methodConnector,
+          session: session,
+          args: args,
+          streamParams: streamParams,
+          message: message,
+          server: server,
+        );
+        break;
     }
+  }
 
-    controller.stream.listen((event) {
-      webSocket.add(event);
-    }, onDone: () async {
-      if (_streamControllers.isEmpty) {
-        await webSocket.close();
-      }
-    });
+  /// Dispatches a message to the correct stream controller.
+  bool dispatchMessage(
+    MethodStreamMessage message,
+    Server server,
+  ) {
+    var streamContext = _inputStreamContexts[_buildStreamKey(
+      endpoint: message.endpoint,
+      method: message.method,
+      parameter: message.parameter,
+      connectionId: message.connectionId,
+    )];
+
+    if (streamContext == null) {
+      return false;
+    }
 
-    _streamControllers[_buildStreamKey(
+    streamContext.controller
+        .add(server.serializationManager.decodeWithType(message.object));
+    return true;
+  }
+
+  void dispatchSerializableException(
+    MethodStreamSerializableException message,
+    Server server,
+  ) {
+    var streamContext = _inputStreamContexts[_buildStreamKey(
       endpoint: message.endpoint,
       method: message.method,
+      parameter: message.parameter,
       connectionId: message.connectionId,
-    )] = controller;
+    )];
+
+    if (streamContext == null) {
+      return;
+    }
+
+    var serializableException =
+        server.serializationManager.decodeWithType(message.object);
+
+    if (serializableException is! SerializableException) {
+      throw Exception(
+        'Expected SerializableException, but got ${serializableException.runtimeType}',
+      );
+    }
+
+    streamContext.controller.addError(serializableException);
   }
 
   String _buildStreamKey({
     required String endpoint,
     required String method,
+    String? parameter,
```

Comment:
Hmm, I don't really see the benefit. Since we are actually missing the parameter, we might as well ommit it completely from the stream key when it is not present.

Or am I missing something we gain here?